### ISSUE TITLE
.value() fonction of constraints (globalconstraint.py)

### ISSUE THREAD

**Issue by 363734** [CPMpy/cpmpy#336]
Some of the constraints return None for this if one of its args is evaluated as None (such as Maximum), some of them check partially if there is a value None (such as Cumulative which checks if one is None, but forgets to flatten the arguments that are list and thus may miss some None elements in it) and some don't do any checks at all (such as Xor) and would simply return an error if this arises. 

Should we uniform this and add full None checks for each of the expression .values() functions?

**Comment by 363734** [comment]
Also, I just notice, by looking at the .value() method of GlobalCardinalityCount constraint, that we could, for all the constraints where the decomposition function is available, do just as the GGC is doing (which calling .value() on the decomposition). If the None is well handled on operators and comparisons, I think for the constraint with a decomposition (not Min, Max, element, count) could just call the evaluation on the decomposition to get the value, since the decomposition use in fine only operators and comparison. This could then be set in the GlobalConstraint class and override for the one without a decomposition

**Comment by JoD** [comment]
What do we mean with `None`? Since we only have total functions, `value()` will always be computable if all variables (leafs in the expression tree) have a value. So `None` covers the case where not all variables may have a value? In that case, it should be properly implemented in all `value()` calls.

Agree with the proposal to have GlobalConstrain provide a default `value()` implementation that works with the decomposition of the constraint. More efficient implementations can then be overridden in the subclass.

**Comment by 363734** [comment]
value() seems to be ok if all the variables are assigned. But if some are not, then there are inconsistencies. 

I just notice that the xor passed the work to a sum operator which indeed checks for the None

but Cumulative does not catch it (because there is no flattening of the lists mapped using argval). And if one of the start is None, then it will try to do None + something. So for me, it is not properly computed as of now. 

And I think refactoring it by evaluating the decomposition will first remove lots of redundant code, and secondly allow to have only basic operations where the None is checked beforehand.

**Comment by IgnaceBleukx** [comment]
Calling the value on the decomposition will not work when that decomposition introduces auxiliary variables.
The new variables will not have a value (as they are not passed through any solve-call) so the `.value()` of the global constraint will always be `None`.
Moreover, creating new expressions in the decomposition, only to check for their value is really inefficient. For example, for the Sudoku Assistant, we first had the `.value()` of the `AllDifferent` constraint to check the pairwise-inequalities for their value. It turned out that during the computation of our explanations, more than 60% of time was spent in the `.value()` function of `AllDifferent`! 
 I think it's safer to just raise an error if `.value()` is not implemented at `GlobalConstraint` level as to me, in general it is not a good idea to rely on the decomposition to get the value of the constraint.

**Comment by 363734** [comment]
and in any case, we need to check that null values are fully checked (either because of the call to simpler operation) or in the .value() fct (which was the initial issue I flagged)

**Comment by IgnaceBleukx** [comment]
Revisiting this as it is still open.
I think we should:
- Throw a NotImplementedError in the superclass when no value() function is implemented
- Add a bunch of tests where we check if any of the argument values is `None` when computing the value of the global. This should propagate to return `None` as the value of the global in the end. I know this is currently failing for many globals. Probably this can be part of the more elaborate testing infrastcture I proposed in #650

**Comment by Dimosts** [comment]
I agree with the first point, but the second one is tricky. 

In general there is the concept that if you have a partial assignment on a subset of variables of the constraints scope, then you can check if this can be extended to any assignment allowed by the constraint. Of course this is often combinatorial and thus would require a solve call, which is not ideal. 

But, even without propagation or solving, we can often determine if a constraint is violated. Assume in alldifferent(x1 - x5), and a partial assignment with x1 = x2. It would be weird to return None in such a case, no?

There are global constraints that we can know that they are not satisfied already from partial assignments in polynomial time. This is formalized as that the global constraint allows partial assignment checking (https://www.duo.uio.no/bitstream/handle/10852/58748/1502.02467.pdf?sequence=1&isAllowed=y).

Polynomial is not ideal still, but we could do it for linear time checking?

I have in mind cases that this would be needed, like CA, where we would want to be able to use the .value() function to simulate the answer of the user in partial assignments, in a problem with global constraints.

**Comment by Dimosts** [comment]
Yes I agree. Lets first make it consistent and try to tighten bounds-computation for boolexpr.
